File: src/types/payment.ts
import { z } from 'zod'

export const PaymentStatus = {
  PAID: 'paid',
  SKIPPED: 'skipped',
  PENDING: 'pending',
} as const

export type PaymentStatusType = (typeof PaymentStatus)[keyof typeof PaymentStatus]

export const paymentSchema = z.object({
  id: z.string(),
  subscription_id: z.string(),
  amount: z.number(),
  paid_at: z.string(),
  due_date: z.string(),
  status: z.enum(['paid', 'skipped', 'pending']).default('paid'),
  notes: z.string().nullable().optional(),
  created_at: z.string().optional(),
})

export type Payment = z.infer<typeof paymentSchema>

export const paymentFormSchema = z.object({
  subscription_id: z.string().min(1, 'Subscription is required'),
  amount: z.number().positive('Amount must be positive'),
  paid_at: z.string().min(1, 'Payment date is required'),
  due_date: z.string().min(1, 'Due date is required'),
  status: z.enum(['paid', 'skipped', 'pending']).default('paid'),
  notes: z.string().optional(),
})

export type PaymentFormData = z.infer<typeof paymentFormSchema>

export interface PaymentWithSubscription extends Payment {
  subscription_name: string
  subscription_color: string | null
  category_name: string | null
  category_color: string | null
}

File: crates/subby-core/src/models/payment.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum PaymentStatus {
    Paid,
    Skipped,
    Pending,
}

impl PaymentStatus {
    pub fn as_str(&self) -> &'static str {
        match self {
            PaymentStatus::Paid => "paid",
            PaymentStatus::Skipped => "skipped",
            PaymentStatus::Pending => "pending",
        }
    }

    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "paid" => Some(PaymentStatus::Paid),
            "skipped" => Some(PaymentStatus::Skipped),
            "pending" => Some(PaymentStatus::Pending),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Payment {
    pub id: String,
    pub subscription_id: String,
    pub amount: f64,
    pub paid_at: String,
    pub due_date: String,
    pub status: PaymentStatus,
    pub notes: Option<String>,
    pub created_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NewPayment {
    pub subscription_id: String,
    pub amount: f64,
    pub paid_at: String,
    pub due_date: String,
    #[serde(default = "default_paid")]
    pub status: PaymentStatus,
    #[serde(default)]
    pub notes: Option<String>,
}

fn default_paid() -> PaymentStatus {
    PaymentStatus::Paid
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct UpdatePayment {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub amount: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub paid_at: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub status: Option<PaymentStatus>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub notes: Option<Option<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PaymentWithSubscription {
    pub id: String,
    pub subscription_id: String,
    pub amount: f64,
    pub paid_at: String,
    pub due_date: String,
    pub status: PaymentStatus,
    pub notes: Option<String>,
    pub created_at: Option<String>,
    pub subscription_name: String,
    pub subscription_color: Option<String>,
    pub category_name: Option<String>,
    pub category_color: Option<String>,
}
